### PR TITLE
#13300: Refactor Llama ROPE op and enable non-power of 2

### PR DESCRIPTION
**Pull request by cglagovichTT**
### Ticket
#13300 

### Problem description
ROPE op was limited to power of 2 sequence lengths. This limitation caused some prefill workloads to pad to much larger lengths than necessary, leading to low perf.

### What's changed
Refactor parallelization scheme for ROPE and enable arbitrary sequence lengths in the ROPE op. 

### Checklist
- [x] Post commit CI passes https://github.com/tenstorrent/tt-metal/actions/runs/11132146506
- [x] T3k Frequent: https://github.com/tenstorrent/tt-metal/actions/runs/11132154225
- [x] Model regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

One last T3k frequent run https://github.com/tenstorrent/tt-metal/actions/runs/11273991047

### CODE REVIEW

File: models/demos/t3000/llama2_70b/tt/llama_generation.py [hunk 1]
```diff
@@ -305,8 +305,16 @@ def _process_logits(self, tt_logits):
...
 def get_padded_prefill_len(seq_len):
-    # Llama supports power of 2 lengths that are greater than 32
-    return max(32, nearest_pow_2(seq_len))
+    """
+    If seq_len is less than 32, pad to 32
+    If seq_len is more than 32, pad to whichever is smaller: a power of 2 or a multiple of 1024
+    TODO: Generalize for max_mm_seq_len different from 1024
+    """
+    pow_2_pad = nearest_pow_2(seq_len)
+    mult_1024_pad = 1024 * math.ceil(pow_2_pad / 1024)
```
Comment (skhorasganiTT) [inline models/demos/t3000/llama2_70b/tt/llama_generation.py:314]: Should this be seq_len/1024 for nearest multiple of 1024?

Comment (cglagovichTT) [reply models/demos/t3000/llama2_70b/tt/llama_generation.py]: Thanks! Took another look, and realized that the padding function would never return 32 even if seq_len <= 32. Should be fixed, please check my work.

Comment (skhorasganiTT) [reply models/demos/t3000/llama2_70b/tt/llama_generation.py]: Looks good, thanks

**Review by skhorasganiTT — APPROVED** [review]

**Review by TT-BrianLiu — APPROVED** [review]
Minor comment, but please fix.

**Review by cfjchu — APPROVED** [review]

**Review by johanna-rock-tt — APPROVED** [review]
